engine/classes/Elgg/Cache/EntityCache.php

```diff
@@ -60,8 +55,25 @@ public function __construct(ElggSession $session, MetadataCache $metadata_cache)
 	public function get($guid) {
 		$guid = (int) $guid;
 
+		if (isset($this->disabled_guids[$guid])) {
+			return false;
+		}
+
 		if (isset($this->entities[$guid])) {
-			return $this->entities[$guid];
+			$entity = $this->entities[$guid];
+		} else {
+			$entity = $this->persisted_cache->load($guid);
+			if ($entity instanceof ElggEntity) {
+				$this->set($entity);
+			}
+		}
+
+		if ($entity instanceof \ElggEntity) {
+			if (!elgg_get_ignore_access() && !has_access_to_entity($entity)) {
```

Comment:
Wouldn't this add a query for every use of the `$this->entities` cache?

Reply:
I'll check this.

Reply:
Yes, but it's all a bit tricky. I'd rather we hit the database with the access SQL rather than just hand out the entity. We could move this to `EntityTable::get()`, check if entity is coming from cache and bail if it's not accessible (but then again how much do we trust `has_access_to_entity()`? What happens if with `enabled` query? Custom SQL injected via `get_sql` hook. We should add some tests for that, but this is an important point you are raising.

Reply:
After seeding, this PR adds 164 queries and ~200ms to /activity.

For me the Q is: Is it safe to assume, that if a user loaded an entity, and nothing access-related has changed, we can provide that entity w/o query. That is, assume the get_sql hook is consistent.

Reply:
> For me the Q is: Is it safe to assume, that if a user loaded an entity, and nothing access-related has changed, we can provide that entity w/o query. That is, assume the get_sql hook is consistent.

Isn't that what the entity cache is for?

Reply:
I am coming back to the problem that ignored access doesn't equal ignored capabilities. We use them interchangeably and that leads to all kinds of problems. A user could load an entity with ignored access for viewing that shouldn't mean they should be able to delete it.

One thing we could do here is add ignored access entity cache, so we don't have to flush caches, but can just swap them depending on the ignored access status.

Reply:
> add ignored access entity cache, so we don't have to flush caches, but can just swap them depending on the ignored access status.

When access is off, just don't write to cache. Reads should be fine right? That's the current system, at least in 2.3

Reply:
Also we should examine where else ignored access is toggled - probably a bunch of places, most of which are unnecessary, that probably flushes the cache on each toggle.

Reply:
That seems right. We only use ignored access for temporary operations